runtime/src/accounts.rs

```diff
@@ -267,144 +268,183 @@ impl Accounts {
+                    }
 
-                        tx_rent += rent;
-                        rent_debits.insert(key, rent, account.lamports());
+                    tx_rent += rent;
+                    rent_debits.insert(key, rent, account.lamports());
 
-                        account
-                    }
+                    account
                 };
 
+                account_found_and_dep_index.push((account_found, account_dep_index));
                 Ok((*key, account))
             })
             .collect::<Result<Vec<_>>>()?;
 
+        if !validated_fee_payer {
+            error_counters.account_not_found += 1;
+            return Err(TransactionError::AccountNotFound);
+        }
+
         // Appends the account_deps at the end of the accounts,
         // this way they can be accessed in a uniform way.
         // At places where only the accounts are needed,
         // the account_deps are truncated using e.g:
         // accounts.iter().take(message.account_keys.len())
         accounts.append(&mut account_deps);
 
-        if validated_fee_payer {
-            let program_indices = message
-                .instructions()
-                .iter()
-                .map(|instruction| {
-                    self.load_executable_accounts(
-                        ancestors,
-                        &mut accounts,
-                        instruction.program_id_index as IndexOfAccount,
-                        error_counters,
-                    )
-                })
-                .collect::<Result<Vec<Vec<IndexOfAccount>>>>()?;
-
-            Ok(LoadedTransaction {
-                accounts,
-                program_indices,
-                rent: tx_rent,
-                rent_debits,
+        let program_indices = message
+            .instructions()
+            .iter()
+            .map(|instruction| {
+                let mut account_indices = Vec::new();
+                let (program_id, program_account) = accounts
+                    .get(instruction.program_id_index as usize)
+                    .ok_or(TransactionError::ProgramAccountNotFound)?;
+                let (account_found, account_dep_index) = account_found_and_dep_index
+                    .get(instruction.program_id_index as usize)
+                    .ok_or(TransactionError::ProgramAccountNotFound)?;
+                if native_loader::check_id(program_id) {
+                    return Ok(Vec::new());
+                }
+                if !account_found {
+                    error_counters.account_not_found += 1;
+                    return Err(TransactionError::ProgramAccountNotFound);
+                }
+                if !program_account.executable() {
+                    error_counters.invalid_program_for_execution += 1;
+                    return Err(TransactionError::InvalidProgramForExecution);
+                }
+                let owner_id = program_account.owner();
+                if !native_loader::check_id(owner_id) {
```

Comment:
Again not familiar enough if call-chain's supported, but would expect `while` instead of `if` here

Reply:
The thing is: CPI of built-in programs is independent of the loader ownership chains here.

For example, [the upgradeable loader (`BPFLoaderUpgradeab1e11111111111111111111111`) does a CPI call to the system instruction processor (`11111111111111111111111111111111`) to create a new account](https://github.com/solana-labs/solana/blob/2c347ac0a5891ecd2137af9ef08b16a550080369/programs/bpf_loader/src/lib.rs#L637). The upgradeable loader however, is owned by the phony native loader (`NativeLoader1111111111111111111111111111111`) which has no entry in the accounts DB and also no owner, thus no relation to the system instruction processor being called.

In other words, I don't understand the purpose of nested ownership for built-ins.

Reply:
Loader chains from top level instructions is lost here, aka no support for non-builti loaders or 3rd party loafers.  Maybe you are bringing up that non-builtin loaders are not supported by cpi?

Reply:
Yes, kind off, we currently have three mechanisms:
- Top-level instructions (with builtin loader ownership chains)
- CPI
- Syscalls

Program runtime v2 is already unifying CPI and syscalls (by modelling syscalls as CPI to built-in programs).
I hope we can get rid of built-in loader ownership chains too and don't see a reason why 3rd party addons / plugins / mods can't just be treated as built-ins as well and simply have a model where all programs can call all other programs using only one mechanism.

Reply:
From what I see during replaying mainnet snapshots, I have not seen a transaction has call chain deeper than 2. Often `BPFLoaderUpgradeab1e` loads one program (and its programdata) accounts. But it doesn't necessary mean removing support of loader call chain will not break existing user programs. 

Is it possible to continue support chain of loader in this refactor PR, so it does not alter function/behavior. Move the change to separate PR instead?

Reply:
Continuing to support it in this is PR might be possible yes. Still, I have not heard any argument why we would keep it, if nobody can show me where it is used. Also, I don't think it is actually possible to construct such a chain in the current program-runtime as one can not change the ownership of a non-empty or executable account.

Reply:
> I don't think it is actually possible to construct such a chain in the current program-runtime as one can not change the ownership of a non-empty or executable account.

If this is the case, then removal of that logic would be good. The simplified code is much cleaner for sure. I don't know enough about it, but I will continue pay attention while I am replaying ledgers. Will post if I find any case of this loader chain is used. Meanwhile, any comments from someone more familiar with it? @jstarry @jackcmay  perhaps?

Reply:
This issue is being discussed in the `virtual-machine` channel on discord.  The current consensus is that we don't need it for move and don't need to retain it here.  If we remove it here, there is nothing in the protocol that I'm aware of that prevents us from adding this functionality back if we decide otherwise later.